File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.py
from collections import deque
import sys

input = sys.stdin.readline


def find(x):
    if parent[x] == x:
        return x
    parent[x] = find(parent[x])
    return parent[x]


def union(a, b):
    x, y = find(a), find(b)
    if x < y:
        parent[y] = x
    else:
        parent[x] = y


def check(x, y):
    #상
    if x - 1 >= 0 and arr[x - 1][y] == 0:
        isRoad, d = False, 0
        for i in range(x - 1, -1, -1):
            if arr[i][y] != 0:
                isRoad = True
                break
            d += 1
        if isRoad and d >= 2:
            if arr[x][y] < arr[x - d - 1][y]:
                paths.add((arr[x][y] - 1, arr[x - d - 1][y] - 1, d))
            else:
                paths.add((arr[x - d - 1][y] - 1, arr[x][y] - 1, d))
    # 하
    if x + 1 < n and arr[x + 1][y] == 0:
        isRoad, d = False, 0
        for i in range(x + 1, n):
            if arr[i][y] != 0:
                isRoad = True
                break
            d += 1
        if isRoad and d >= 2:
            if arr[x][y] < arr[x + d + 1][y]:
                paths.add((arr[x][y] - 1, arr[x + d + 1][y] - 1, d))
            else:
                paths.add((arr[x + d + 1][y] - 1, arr[x][y] - 1, d))

    # 좌
    if y - 1 >= 0 and arr[x][y - 1] == 0:
        isRoad, d = False, 0
        for i in range(y - 1, -1, -1):
            if arr[x][i] != 0:
                isRoad = True
                break
            d += 1
        if isRoad and d >= 2:
            if arr[x][y] < arr[x][y - d - 1]:
                paths.add((arr[x][y] - 1, arr[x][y - d - 1] - 1, d))
            else:
                paths.add((arr[x][y - d - 1] - 1, arr[x][y] - 1, d))

    # 우
    if y + 1 < m and arr[x][y + 1] == 0:
        isRoad, d = False, 0
        for i in range(y + 1, m):
            if arr[x][i] != 0:
                isRoad = True
                break
            d += 1
        if isRoad and d >= 2:
            if arr[x][y] < arr[x][y + d + 1]:
                paths.add((arr[x][y] - 1, arr[x][y + d + 1] - 1, d))
            else:
                paths.add((arr[x][y + d + 1] - 1, arr[x][y] - 1, d))

def bfs(a, b):
    q = deque([(a, b)])
    arr[a][b] = node
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and arr[nx][ny] == 1:
                arr[nx][ny] = node
                q.append((nx, ny))


n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]
node = 2
paths = set()
for i in range(n):
    for j in range(m):
        if arr[i][j] == 1:
            bfs(i, j)
            node += 1

for i in range(n):
    for j in range(m):
        if arr[i][j] != 0:
            check(i, j)

paths = sorted(paths, key=lambda x:(x[2]))


parent = [i for i in range(node - 1)]
result = 0
for a, b, cost in paths:
    if find(a) != find(b):
        union(a, b)
        result += cost

for i in range(1, node - 1):
    parent[i] = find(i)

if len(set(parent[1:])) != 1:
    print(-1)
else:
    print(result)

File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.java
import java.util.*;
import java.io.*;
public class Main {
	static int n, m, cnt, answer;
	static int[] parent;
	static int[][] arr;
	static int[] dx = {-1, 1, 0, 0};
	static int[] dy = {0, 0, -1, 1};
	static List<Node> graph;

	static class Node implements Comparable<Node>{
		int a, b, w;

		public Node(int a, int b, int w) {
			this.a = a;
			this.b = b;
			this.w = w;
		}

		@Override
		public int compareTo(Node o){
			return this.w - o.w;
		}
	}
	public static void islandSet(int x, int y, int num){
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i], ny = y + dy[i];
			if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
			if (arr[nx][ny] == -1){
				arr[nx][ny] = num;
				islandSet(nx, ny, num);
			}
		}
	}

	public static void makeBridge(int x, int y){
		for (int k = 0; k < 4; k++) {
			int nx = x, ny = y, len = 0;
			while(true){
				nx += dx[k];
				ny += dy[k];
				if (nx < 0 || nx >= n || ny < 0 || ny >= m || arr[nx][ny] == arr[x][y]) {
					break;
				}
				if (arr[nx][ny] != 0){
					if (len >= 2) {
						graph.add(new Node(arr[x][y], arr[nx][ny], len));
					}
					break;
				}
				len ++;
			}
		}
	}

	public static int find(int x){
		if (parent[x] == x) return x;
		return parent[x] = find(parent[x]);
	}

	public static boolean union(int x, int y){
		x = find(x);
		y = find(y);
		if (x == y) return false;
		if (x < y) parent[y] = x;
		else parent[x] = y;
		return true;
	}

	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[n][m];
		graph = new ArrayList<>();

		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < m; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
				if (arr[i][j] == 1) arr[i][j] = -1;
			}
		}

		int num = 1;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (arr[i][j] == -1){
					arr[i][j] = num;
					islandSet(i, j, num++);
				}
			}
		}

		parent = new int[num];
		for (int i = 0; i < num; i++)
			parent[i] = i;

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (arr[i][j] != 0)
					makeBridge(i, j);
			}
		}

		Collections.sort(graph);
		cnt = 0;
		answer = 0;
		for (Node node : graph) {
			if (union(node.a, node.b)){
				answer += node.w;
				if (++cnt == num - 1)
					break;
			}
		}

		System.out.println((cnt == num - 2) ? answer : -1);
	}
}